### ISSUE TITLE
[MU4 Issue] The user should be able to unmute hidden instruments

### ISSUE THREAD

**Issue by DmitryArefiev** [musescore/MuseScore#10951]
**Describe the bug**
See #9357

If you hide an instrument using the 'Instruments' panel, the instrument should be muted in the mixer. 
The user **should be able to unmute** it if they want

https://user-images.githubusercontent.com/90187801/160684542-0789944b-9c0a-4f97-a989-697c02ee5476.mp4

**Expected result**
You should give user ability to unmute hidden instruments in Mixer 
(or implement something in Instruments panel - see https://github.com/musescore/MuseScore/issues/9357#issuecomment-950046752 )

**Platform information**
macOS, Windows, Linux

**Comment by MarcSabatella** [comment]
This isn't a feature request, it's a regression - it has always worked in the past and lots of people rely on it as a way to get all sorts playback effects you can't otherwise (rhythm section parts notated with just slashes, various ornaments and other notations not supported for playback directly.  I have verified this with one of my big band scores; no drum playback.

The good news is, it seems you *can* still access the mute controls, if instead of hiding the *instrument*, you hide its *staff* (or staves).  So it seems the functionality is still there (thankfully!) - it just for some reason isn't working as it should when hiding the full instrument instead of its staff/staves.

It's still a regression in that it means any imported 3.x score using invisible playback staves won't play correctly without employing a hard-to-discover workaround.  And even for new scores, it isn't necessarily obvious that you even can hide the staves individually or that this would work when hiding the instrument doesn't.

**Comment by Tantacrul** [comment]
Never mind. I found it.

**Comment by Tantacrul** [comment]
@bkunda - would be good to perhaps consider a design here. Perhaps just the ability to toggle on/off the sound in the mixer?

**Comment by bkunda** [comment]
> @bkunda - would be good to perhaps consider a design here. Perhaps just the ability to toggle on/off the sound in the mixer? Perhaps something in the Instruments panel itself?

Do we really need a design for this? (Maybe I'm missing something).

I think at this stage our goal should simply be to restore the functionality as it was in MS3.6, which would mean: 

1. Not automatically muting an instrument that has its stave(s) hidden in the instruments panel
2. Allowing the user the ability to mute and unmute any hidden instrument in the mixer

This behaviour presupposes an expectation that what you see on the score is not necessarily always linked to playback – i.e. that the two are separable, allowing users to create playback effects that are not directly connected to what is on the score. This might of course be a moot point, but I'd suggest that any discussion about this would be better reserved for an MS4.x release.

**Comment by Tantacrul** [comment]
I don't think we should return to 3.6 behaviour. It is more weird for the sounds to persist when the instrument it hidden.

It could just be that we allow users to unmute in the mixer. However, this might be a hacky solution, so I thought it deserved more thought then that. Perhaps we can use this solution for 4.0 but the logic would need to be spelled out:

- If you hide an instrument, it would automatically mute the instrument in the mixer
- If you unmute, the instrument can now be heard again
- (Part I'm not sure about) If you show and then hide the instrument again, it would still be heard (and users can mute it again if they want)
- (OR) every time you hide an instrument, it always mutes in the mixer

It is also worth exporting a score from 3.6 with a hidden staff and importing it to Audio.com to see what happens currently.

We may need to support this in 4.0. Not sure.

**Comment by bkunda** [comment]
OK I think I can be persuaded to this (I have storied-out some user cases in order to think this through, which I'd be happy to share elsewhere if it helps discussion). 

In this case, my suggestion would be that we adopt a fairly hard line about the interaction between showing/hiding instruments and their playback property, i.e. unmuted/muted. This is in order to keep the behaviour predictable. 

So, by default:

- When an instrument is hidden in the instruments panel, its playback is automatically muted
- When an instrument is not hidden (or rather, shown) in the instruments panel, its playback is by default un-muted

In either case, this playback property can be overridden in the mixer:

- A hidden instrument can be unmuted (currently broken)
- An unhidden instrument can be muted (existing functionality) 

If the user changes the visibility state of the instrument in the instruments panel, playback for that instrument reverts to its default state. 

For example:

- The user un-mutes a hidden instrument to hear its playback. They then show the instrument in the instruments panel – playback continues. They then hide the same instrument again – playback is muted. 

What do you think @Tantacrul? Perhaps we can also chat separately about importing to audio.com. 

As for supporting this in MS4.0, I think we definitely need to address the issue of not being able to un-mute a hidden instrument, so there's _only_ (he says, naïvely) one more step to establish beyond that, which is to ensure that the instrument becomes muted again if the user shows and hides it in the instruments panel. Not implementing this extra step in MS4.0 could cause some unpredictable behaviour, with playback occurring by default for hidden instruments in some cases but not others (my spider sense says this could be problematic).

**Comment by MarcSabatella** [comment]
Makes sense.  But either way, the regression where MU3 scores will incorrectly needs to be solved.

**Comment by MarcSabatella** [comment]
Reported again in https://github.com/musescore/MuseScore/issues/10951.  I had forgotten about this.  Hiding instruments but still hearing them is pretty essential for jazz scores where people use this for rhythm section parts or improvised solo sections that are notated just with slashes.  Also any number of places where people wish to write out a complex ornament playback etc.

I still say, not once during the 10+ years of MuseScore 1, 2, or 3 can I recall anyone ever complaining that invisible instruments could be heard by default.  To me it's expected behavior.  The simple solution for 4.0 is to just remove the code disabling the mute button, and then it can always be revisited.

**Comment by MarcSabatella** [comment]
I went ahead and created a PR for this (https://github.com/musescore/MuseScore/pull/13282; ignore my local WIP version), simply restoring the MU3 behavior where hiding an instrument and muting it are independent.  It's simple and it's what people are used to.

I'm assuming that at this point, user research and a new design for an alternative scheme is not in the cards for 4.0, which is fine by me.  I'm also fine with seeing this revisited in the future.  But meanwhile, the regression needs to be for fixed for 4.0, as there will be tons of scores that suddenly don't play correctly, with no possible way to fix them.  Plus it's an important capability for new scores that won't have any workaround otherwise.

**Comment by Tantacrul** [comment]
I think we have to stick with the current behaviour and add the ability to unmute instruments - as suggested by Dmitri. I do not think it is simple or intuitive to hear music you don't see by default.

It would also require a large inconsistency with part scores (do we hear all other, invisible instruments when we play a part score by default?).

**Comment by MarcSabatella** [comment]
As someone who uses invisible playback staves almost every day (as do almost all other jazz arrangers), I guess I don’t see why you’d think of this as less intuitive than *seeing* staves you don’t *hear*.  And yet no one seems to be suggest we automatically *hide* staves when *muting* them.  Why is it intuitive in one direction but not the other?

But as long as the regression is fixed - that existing scores that uses this technique  continue to play correctly without workarounds - I’m sure people can adapt to needing a few more steps to accomplish this on new ones.  Just seems like a weird thing to ask everyone to have to do.

**Comment by MarcSabatella** [comment]
Just curious - do all other major notation programs automatically mute staves when hiding them, but not the reverse?  If so, the. I can see why it might seems intuitive for people used to those programs.

**Comment by Tantacrul** [comment]
You might want to mute tracks so you can momentarily isolate others. It's not really a comparison I'd draw.

You've explained why you want to hide sheet music but still want to hear it, so we accept a user should be able to unmute a hidden track. However, making it work like that by default doesn't make sense.

**Comment by cbjeukendrup** [comment]
@Tantacrul Allowing the user to unmute tracks for hidden instruments is not straight-forward at all, especially when it comes to part scores.

The problem is that mute/solo state is in principle a project-wide setting. Currently, the logic to determine whether a track is _actually_ muted is as follows. A track is _actually_ muted if either:
1. the track has been muted by the user,
2. or a different track has been soloed by the user,
3. or the staff corresponding to that track is not present/visible in the currently viewed part.

If a track has not been muted by the user but is still actually muted for reason 2 or 3, the "mute" button in the Mixer is disabled, because it wouldn't do anything useful: the track will stay muted because of reason 2 or 3.

Note that the user's chosen mute state is never changed by the playback system itself. The user's choice is also what gets saved into the file; so if a track was automatically/temporarily muted for reason 2 or 3, this will not be stored in the file but rather computed dynamically from other data.

If we want to allow the user to unmute tracks that would be muted for reason 2 or 3, we have to use a different logic, but it is not clear to me how it should work exactly. Some things that make it difficult are:
- we should allow overriding reason 3, but not reason 2;
- mute/solo state is (currently) a project-wide setting and can't be stored per-part;
- it's unclear how/what should be saved in the file (for example, if the user switches to a part score, then we would automatically mute the invisible staves, but that should _not_ be saved in the file, because otherwise, next time the user opens the score, those tracks will be muted in the full score too).

So what we need to solve this issue, is a clear definition of how the new logic should work. What would be your suggestion?

One suggestion of mine:
instead of an on/off setting, we can make "mute" a three-state property:
- unmuted -> user has specified that this track should always sound
- default -> muted-ness of this track will be determined by the visibility of the staff in the currently viewed part
- muted -> user has specified that this track should never sound

(reason 2 about the "solo" setting will always take priority over whatever choice the user has made)

This could be how it works internally, but I've no idea how we should represent it to the user, because of course from UX POV a "mute" button should look like a simple on/off switch.

**Comment by Tantacrul** [comment]
I think the way we'd want it to work (just side-stepping the technical considerations for a moment) is this:

1. A user can un-mute a hidden instrument using the mixer (this would mean we'd have to create additional logic for hidden instruments)
2. If another instrument is soloed, then the mute buttons are disabled, so the user can't unmute anything until the solo instances are cancelled
3. I don't think we should create a third state in the UI because this isn't a problem users will understand.

If it is too complex for 4.0, then we wait until 4.x to do it. We unfortunately can't perfect everything at this late stage and I'd rather hear from some users if this is really a feature they consider to be important before committing a lot of time to 'solving' it.

**Comment by Tantacrul** [comment]
I'm also open to the idea of adding UI to the instruments panel to allow for turning on/off sound for each. Since this requires an actual redesign, it can't make it in for 4.0.

**Comment by henkdegroot** [comment]
> We unfortunately can't perfect everything at this late stage and I'd rather hear from some users if this is really a feature they consider to be important before committing a lot of time to 'solving' it.

If you want more users "opinions" you should make this issue more visible (i.e. a post in the forum and get feedback). It is a feature I use often in MS3. My guess is that users will experience "playback" issues after open a MS3 score in MS4 and consider there is an issue with migration scores from MS3 to MS4 and may (as a result) be reluctant to start using MS4.

**Comment by Tantacrul** [comment]
btw, Marc: I do get why you want this solved pronto. I was hoping there'd be a quick fix here... but it's worth pointing out that the new setup is quite different to MS3 and sets up different expectations:

- The instruments panel doubles as the way users manage the visibility of their parts. So if it works one way there, it should work the same way on the main score.
- Anyone using the instruments feature to create alternate passages or as a method of displaying combined instruments (eg: 4 horns on one stave) in the main score, while having each instrument separated for parts will end up hearing loads of things they don't want to hear.

---

One last point, which I don't really make enough: for an app like this, my general approach is that **UX is everything**. I'd rather delay a feature in order to get the UX right rather than provide a quick fix solution that ostensibly solves the problem but in a confusing way, which then needs to be undone later.

**Comment by Tantacrul** [comment]
@henkdegroot

> If you want more users "opinions" you should make this issue more visible 

We've agreed to come up with a solution, the only remaining question is when we do it. We unfortunately, cannot delay this release any further.

**Comment by MarcSabatella** [comment]
I'm coming around to being able to let this go, except for the matter of existing scores.

While I remain unconvinced that the new design will automatically change people's expectations about the connection between hiding and muting, the fact that MU4 makes it possible to hide staves without hiding the instruments is, I think, our ace in the hole here.  It means you no longer even need to add a whole new instrument to get an invisible playback staff for the usual sorts of purposes people use these for currently.  Instead, you can just add a new staff to an existing instrument.  So going forward, it shouldn't be a big deal to re-train people to think this way, and it may even seem like a win once people get used to it.

But, my main concern remains existing scores. It is still a pretty huge regression that tons of existing scores won't play back correctly without some action being taken here.  Since it is the case that invisible *staves* are still heard by default, and the ability to hide staves as opposed to instruments was not present in MU3 and earlier, my proposal would be, for older scores, we automatically convert invisible instruments into invisible staves on import.  That is, still add the instrument, but don't mark the instrument invisible - mark its staves invisible.

That should fix the playback of existing scores without affecting anything about the current UX, and it will also help people see how this is done,

**Comment by MarcSabatella** [comment]
New PR https://github.com/musescore/MuseScore/pull/13293 implements the import behavior described above.  Tested for 3.x and 2.x; I don't have any 1.x scores with invisible playback staves handy to check.

**Comment by MRT-CH** [comment]
I have more than 100 scores that use the 3.6 behavior. I will not move them to 4.0 as long this is not fixed.

**Comment by Jojo-Schmitz** [comment]
I frequently add staves and make them invisible for one single purpose: playback. 
Use case: mixed choir in unison, women visible (and with an optional ottava bassa clef), men hidden (and an octave lower of course). Only reason for having those is for playback.

**Comment by MarcSabatella** [comment]
I'd like to suggest that my PR https://github.com/musescore/MuseScore/pull/13293 be merged for Beta if at all possible.  It makes no changes whatsoever to the UX, no changes whatsoever to any new scores created in MU4.  It simply makes sure that existing MU3 scores with invisible instruments play back correctly, by taking advantage of the way staff visibility works now (converting invisible *instruments* into invisible *staves*).  There are a number of other issues that will interfere with proper playback of existing scores, and many of those we can't easily do anything about, but this one we can and should.

Long term, barring additional user feedback, I'm fine with the existing UX where invisible *instruments* are muted but invisible *staves* are not.  However, another significant regression is that there is no longer a way to mute one staff of a grand staff but not the other.  So, I would like to observe that adding a mute control to the instruments panel - say, an ear icon next to the existing eye icon - would solve that problem nicely, and *also* partially address concerns about the mixer being too big.  So, definitely something to consider for 4.x.

**Comment by Tantacrul** [comment]
We're going to allow users to mute / unmute hidden instruments for sure. We're going to do quite a lot of cool playback stuff with the instruments panel in fact. However, the dev team have advised that this is a complex thing to try and rush in (at the time, we were discussing it for alpha 1 or alpha 2) 

But we can't have an inconsistent system where some scores work one way and other scores work another way. It's just something we'll have to build into MS4 'properly', if you know what I mean?

**Comment by MarcSabatella** [comment]
I still think there is misunderstanding about this.  Hiding a single staff within an instrument works *perfectly* right now in MU4.  It's fine as is and will continue to be fine even if new cool features are *also* added.  Adding a second invisible staff to an existing instrument works is exactly how people *should* be adding invisible playback notes for most use cases.  Like, for instance, creating an invisible second staff for VST key switches.  Or creating a second drum staff for playback while using slash notation for the notation.  No changes at all are required to make that work in MU4 - it already does, and works well.  So that is what people will start doing in MU4 from day one.

I'm just asking that we allow existing MU3 scores to take advantage of this feature automatically - so they will work the same way as MU4 scores.  That is all my code does.  It doesn't make them work *differently* from MU4 - it makes them work *the same*.

**Comment by MRT-CH** [comment]
I am not sure whether I understood what your code exactly doaes Marc - 
and as IT professional with over 40 years in that field I understand and support the case that a system design should be logical and  consistent. Otherwise you will regret this sooner or later.
I also want to support the case posted by Jojo : Choir scores and playback audio tracks for practice. 
The background are some common practices in choral notation and some of the current limitations of Musescore (at least the 3.6 Version)
- Common choral notations often tries to be compact - this is sometimes not always constistent when you think about playback.
In a choral score one choir voices i.e alto (this is not a Musescore voice !) sometimes split in two (A1 and A2). The common parts might be written in one notation/Musescore voice with the text "unison". When voices split they might be writen as two voices in one staff or two separate stafflines - all in one score . 
Now for practice an Alto 2 would like a playback track that plays her voice from start to end either solo or louder than all other choral parts. I know hundreds of singers that work that way. 
Since Musescore 3.6 playback is not able to control dynamics on individual voices and mixer control for voices is only possible with a awkward manipulation of the mscx file - I used a workaround with doubled stafflines. The familiar notation ones you see and the hidden ones you hear and can adjust with volume control.  
This is not elegant either but it works and has been used for more than 400 choral scores. 
But if some of the necessary features are not implemented (i.e. dynamics and volume control for voices) and  the workaround path gets blocked this is really dissapointing. I am hoping for a reasonable solution - not rightaway - although a will miss some of the new 4.0 features but within a some of the next updates.  
As example a link to a score that can plays six voices SSATBB in playback end to end (set solo or louder than the rest) while maintaining the the compact visible notation of the original engraving. Can't attach musescore mscz here :-(
[https://musescore.com/user/1007791/scores/6187770/s/oI7d6A?share=copy_link](url)

**Comment by jeetee** [comment]
This is one of the issues that came up a few times in the forums already. The desire to hide instruments and keep them unmuted for playback purposes is definitely there.

**Comment by skalleAnka** [comment]
This has completely removed a use case I think should be fairly common, at least for users who want to migrate from Guitar Pro.

When you are practicing an instrument, you generally want to see the notation for one instrument/part and hear all the other instruments as backing tracks. The very opposite of what is happening now.

I would probably be all right if I could just unmute tracks, but I don't really find it logical at all that hiding a part should also mute. The eye icon in a way becomes an implied ear also. What would feel natural is that the eye controls visibility only and changes to playback mute/solo are always done in the mixer.

**Comment by Erisceres** [comment]
For use cases, I rely on this feature for Baroque scores where the basso continuo staff covers the entire continuo section (violoncellos, contrabasses, harpsichords, archlutes, etc.). I want the final score to appear authentic by only utilising a staff that covers a bass line with figured bass while hiding the actual instruments but hearing their interpretation in playback.

Similarly, I also work with jazz lead sheets where you have the basic melody and chord information in a single staff, but hide the actual instruments with which I want to interpret the music, hearing them in playback.

**Comment by Erisceres** [comment]
Ah, I understand now. It took me a little while to figure out how to hide staves and not the instruments, but everything seems to be working fine now. Thank you.

**Comment by oMrSmith** [comment]
I agree with all Marc said and think: Software should not shape the user.
Also: I find that it has been a strengh of MuseScore that labels match their result.
If something says "hide", why should it mute?

**Comment by scorster** [comment]
The following position has been expressed: The current hide/mute behavior is to be preserved and therefore we wait for some complicated wiring to allow unmuting the hidden in the mixer.

This reminds me of Ptolemaic vs Copernican views of the solar system.

##A summary of some objections posted here:

**oMrSmith** wrote > 
_If something says "hide", why should it mute?_

**Marc Sabatella** wrote >
_Hiding instruments but still hearing them is pretty essential for jazz scores where people use this for rhythm section parts ..._ 


**Marc Sabatella** wrote >
_... not once during the 10+ years of MuseScore 1, 2, or 3 can I recall anyone ever complaining that invisible instruments could be heard by default. To me it's expected behavior. The simple solution for 4.0 is to just remove the code disabling the mute button, and then it can always be revisited._

[Comment] And I'll note, we've had that option all along in MuseScore 3 by unchecking instruments in the Instrument panel. And no one ever complained

**skalleAnk** wrote >

_I don't really find it logical at all that hiding a part should also mute. The eye icon in a way becomes an implied ear also. What would feel natural is that the eye controls visibility only and changes to playback mute/solo are always done in the mixer._

## And there's dialog on the MuseScore.org forum

jeetee:

_This [issue] came up a few times in the forums already. The desire to hide instruments and keep them unmuted for playback purposes is definitely there._

Here are some of my thoughts on "eye" issues not limited to the binding of hide/mute:

https://musescore.org/en/node/339171#comment-1158760

Thanks for considering!

scorster

**Comment by cbjeukendrup** [comment]
It is not very constructive to keep filling this thread with comments **why** we should change this and how many people want this. We already know that. Making an issue unwieldy with many comments that don't add information, makes it only less likely to get fixed, so you'd better watch out 😄

The question that still has not been answered is **how** it should work. What should be the interaction between manual and automatic muting? Maybe not everyone realises that this problem is more complicated than it seems, because the behaviour w.r.t. part scores also needs to be sensible.

The reason that tracks for invisible instruments are muted, is part scores. In a part score, it is a reasonable default that only the instrument belonging to that part is audible. Therefore, we need to mute the tracks for non-included or invisible instruments.

To turn this conversation in a constructive direction, I'll list some options:
1. The current situation
    - The manual solo/mute states are stored file-wide, so those are the same in the full score and in the part scores. (If you mute a track in one part score, it will be muted too in all other part score and the full score, and vice versa.)
    - MuseScore never automatically modifies the manual solo/mute states; only the user can change them.
    - When calculating whether a track should actually be muted in the current situation, we first apply the manual solo/mute states, and then on top of that, we mute some tracks based on the instruments that are included in the current part score.
    - The advantage of this way is that MuseScore always knows what has been dictated by the user and what it has done itself, so changes made by the user are never lost.
    - The disadvantage is that the user doesn't have control over what MuseScore does itself, which is exactly what this issue is about.
1. Not applying any automatic behaviour at all
    - The manual solo/mute states are still stored file-wide, so those are the same in the full score and in the part scores.
    - MuseScore never automatically mutes tracks. So in the full score, you will by default hear all instruments, even invisible ones, and in part scores, the same; you will hear all instruments of the full score, even invisible ones. And still, when muting a track in one part score, it will be muted in all other parts and in the full score too.
1. Not applying any automatic behaviour at all, and storing states per part score
    - Rather than file-wide, the manual solo/mute states are now stored for each part individually. So there is no undesired (but also no desired) synchronisation anymore. 
    - Because states are now per-part, we can come up with a smart initial state, for example, in the full score, all instruments are heard by default, and in part scores, only the initially visible instruments are heard. 
    - But if you make more instruments visible in a part score, MuseScore will not automatically unmute their tracks for that part.
1. Letting MuseScore change the manual solo/mute states so that the user can change them back
    - The manual solo/mute states are stored file-wide, so those are the same in the full score and in the part scores.
    - When you switch to a different part score, or when you hide an instrument, MuseScore will change the manual solo/mute state.
    - Advantage: the user can change it back.
    - Disadvantage: MuseScore's changes are "destructive", to use a technical term. That means that MuseScore's changes as well as your changes are immediately definitive, and you can't know whether a certain change was made by the user or by MuseScore. So therefore every time you switch to a different part, MuseScore will do its own thing, overriding whatever the user has done.
1. Letting MuseScore change the manual solo/mute staves, but only within a part score
    - This is a compromis between the previous two options.
    - Rather than file-wide, the manual solo/mute states are stored for each part individually. So there is no undesired (but also no desired) synchronisation anymore. 
    - We come up with a smart initial state; for example, in the full score, all instruments are heard by default, and in part scores, only the initially visible instruments are heard. 
    - When the user makes more instruments visible or invisible in the current part, the corresponding tracks are automatically muted. 
        - This is a "destructive" change again, but this is less bad because it's only a local one, confined to the current part score, so less work of the user is lost, if any. 
        - Also, it would be an "incremental" change only: instead of completely recalculating the solo/mute states, only the instrument that is being added or removed will be unmuted/muted. The user's changes to other tracks will be preserved.
        - (If we would make a non-destructive change instead, i.e. applying some logic on top of what the user has done, we're pretty much back in the current situation again.)

So, which one should we choose? Or a combination of one of these? Or yet another one?

Once this question is finally answered, we can start implementing the chosen solution.

**Comment by MarcSabatella** [comment]
That's a lot of options, and while I have *some* understanding of the internals, it's not necessarily enough to have a really good feel for what makes sense internally.  So I'm going to answer a little more generally, and then see if we can map that to some of the specific internals.

First, I would observe there is an important flip side to this - people *also* want to be able to view a part and yet still hear the full score.  This was possible via a single checkbox in the mixer in MU3.  Not by default of course, but ideally it should be that easy.  I'll come back to this below.

I'm also fine with the idea that *by default*, hiding an instrument might mute it in the score, as long as there is an *obvious and easy* way to not do that.  This to me doesn't mean hiding it in one window then unmuting it in another window entirely - it means, all controls in one place.  So, for example, the eye icon in the Instrument panel could be accompanied by an ear icon, and the behavior of toggling the eye could automatically set the ear to the same state, but the ear icon would be independent.  Huge side win: people wanting to be able to to mute things without opening the mixer.  Even bigger win if there were *also* a solo button in the Instruments panel.  Even if these don't happen right away, it's worth looking ahead in designing this.

Of course, if I enable the option I mentioned above to still hear the full score while viewing a part, I don't expect to need to go through each instrument to decide what to hear and what not - I want this to literally mean, play whatever I would hear if I were actually viewing the score.  So this option would literally just substitute the full score playback for the part playback.

Hmm, what about explicitly unmuting some hidden instruments in the part but not others?  This seems like a "nice to have" that people could probably live without if it complicates things.

I *suspect* there is a relatively straightforward way to achieve all this, and it could potentially be doable *without* needing to store mute states per part, especially if the mute can be multi-state (eg, muted, unmuted, and auto).  But those are implementation details I don't have a great handle on.

**Comment by oMrSmith** [comment]
Merry Christmas, everybody!

**Comment by fjacomme** [comment]
Hi,
I was often using MuseScore3 to practice my own part along the other instruments as backing track, so I am interested in this change.

>  especially if the mute can be multi-state

@cbjeukendrup already seemed aware of this need with this TODO !
https://github.com/musescore/MuseScore/blob/002fd32935d86e54ac630f75b12f0e7016c948f9/src/playback/qml/MuseScore/Playback/internal/MixerMuteAndSoloSection.qml#L53

I tried the fix in [#13282](https://github.com/musescore/MuseScore/pull/13282) and with an "enabled: true" in the qml it works well enough for my use case.
 
Another solution could be to remove the automute algorithm when changing visibility and have it run manually with some "hear only the instruments not in this part" and "hear only the instruments in this parts" buttons.

For reference in Ableton (without saying MuseScore should copy the behaviour):
* collapsing a track does not mute it
* soloing a track still leaves every mute buttons active
* soloing a muted track unmute it temporarily
So it seems like a good pratice to never "disable" the button and let the user have control.

**Comment by oMrSmith** [comment]
Even if hidden instruments can be unmuted in the future,
they should not be muted automatically within individual parts.

Just think about it: If you want to proof read an individual part
and still hear the rest of the score, you would have to 
unmute each other part in the mixer. This would be ridiculous.

**Comment by MarcSabatella** [comment]
Right now everything single instrument other than the one being shown already is hidden, but muted.  That's kind of the whole point - normally you *don't* want to hear all those hidden instruments when playing a part.  Unless you want to *always* hear the whole score, the automatic muting absolutely makes sense in parts - otherwise you'd always have to manually mute everything just to hear the part normally.  What is needed is a way to bypass that normal auto-muting for the few cases where you *do* want to hear the score while viewing a part.  A simple checkbox like MU3 had would do the trick, as discussed above.

**Comment by oMrSmith** [comment]
> normally you _don't_ want to hear all those hidden instruments when playing a part

I think, I do, why not?

> otherwise you'd always have to manually mute everything just to hear the part normally.

I would just hit the "solo" button.

**Comment by MarcSabatella** [comment]
I think for most people, having to hit the solo button every single time they want to hear a part would be a major step backwards.  The vast majority of time that people play a part, they want to hear the part only.

**Comment by oMrSmith** [comment]
Wouldn't it make sense then
to simply "solo" visable instruments 
within individual parts by default
instead of locking the mixer.

**Comment by MarcSabatella** [comment]
An option to allow one to play the entire score definitely makes sense, no one has disagreed with that.
As for changing the overall behavior of the relationship between hiding and muting and the function of the solo button, please read through the entire thread to understand all the different cases that need to be considered.

**Comment by MRT-CH** [comment]
Sure , when parts are created via the menu, then all other parts should be initially hidden AND muted. But this can be done by two actions in the procedure and does not require the combination that hiding auto-mutes a part. If a user shows or hides a part explicitly through the UI this should not have any side effect on the playback. The desired functionality should lead the implementation and not the other way round. This might also require small modification of the code for the creation of the parts.

**Comment by bkunda** [comment]
Hi all,
Thanks for all of the input on this issue. There are a lot of important use cases we need to consider here, and it's worth reiterating that the way we ultimately tackle this issue will need to take into account implications for part scores, and potentially also the instruments panel. 
To this end, we think it will be most useful to convert this issue to a discussion for now because it's not something we can solve straight away. The internal team also has a lot to discuss at the technical level; we want to make sure our solution is the best one for score–part interaction, while also taking into account future design ideas we have for both the instruments panel and the mixer.
